fenicsadapter/fenicsadapter.py

```diff
@@ -1,759 +1,458 @@
-            self._write_data = None  # a numpy 1D array with the values like it is used by precice (The 2D-format of values is (d0x, d0y, d1x, d1y, ..., dnx, dny) The 3D-format of values is (d0x, d0y, d0z, d1x, d1y, d1z, ..., dnx, dny, dnz))
-            self._write_function_type = None  # stores whether write function is scalar or vector valued
-
-        # read data related quantities (read data is read by this solver from preCICE)
-        self._read_data_name = self._config.get_read_data_name()
-        self._read_data_id = self._interface.get_data_id(self._read_data_name, self._mesh_id)
-        self._read_data = None  # a numpy 1D array with the values like it is used by precice (The 2D-format of values is (d0x, d0y, d1x, d1y, ..., dnx, dny) The 3D-format of values is (d0x, d0y, d0z, d1x, d1y, d1z, ..., dnx, dny, dnz))
-        self._read_function_type = None  # stores whether read function is scalar or vector valued
-
-        # numerics
-        self._precice_tau = None
-        self._my_expression = interpolation_strategy
+            raise Exception("Rank of function space is neither 0 nor 1")
 
-        # checkpointing
-        self._checkpoint = Checkpoint()
+        return read_data
 
-        # function space
-        self._function_space = None
-        self._dss = None  # measure for boundary integral
-
-        # Nodes with Dirichlet and Force-boundary
-        self._Dirichlet_Boundary = None  # stores a dirichlet boundary (if provided)
-        self._has_force_boundary = None  # stores whether force_boundary exists
-        
-    def _convert_fenics_to_precice(self, data):
-        """Converts FEniCS data of type dolfin.Function into Numpy array for all x and y coordinates on the boundary.
-
-        :param data: FEniCS boundary function
-        :raise Exception: if type of data cannot be handled
-        :return: array of FEniCS function values at each point on the boundary
-        """
-        if type(data) is dolfin.Function:
-            x_all, y_all = self._extract_coupling_boundary_coordinates()
-            return np.array([data(x, y) for x, y in zip(x_all, y_all)])
-        else:
-            raise Exception("Cannot handle data type %s" % type(data))
-            
-    def _write_block_data(self):
-        """ Writes data to preCICE. Depending on the dimensions of the simulation (2D-3D Coupling, 2D-2D coupling or
-        Scalar/Vector write function) write_data is first converted.
+    def write_data(self, write_function):
         """
+        Writes data to preCICE. Depending on the dimensions of the simulation (2D-3D Coupling, 2D-2D coupling or
+        Scalar/Vector write function) write_data is first converted into a format needed for preCICE.
 
-        assert (self._write_function_type in list(FunctionType))
+        Parameters
+        ----------
+        write_function : FEniCS function
+            A FEniCS function consisting of the data which this participant will write to preCICE in every time step.
```

Comment:
I'm not sure whether "FEniCS function" is the correct term (I know, I use it a lot myself). Is [this here](https://fenicsproject.org/docs/dolfin/1.4.0/python/programmers-reference/functions/function/Function.html) the correct function we are talking about? You could write something like `dolfin.functions.function.Function`, then everything is clear.